### DISCUSSION TITLE
Various quality of life features

### DISCUSSION THREAD

**Discussion by dylanrafael05** [dotnet/csharplang#4714] [Language Ideas]
Lightning Round!

Allow `not x` in place of `x is not null`:
```cs
if(not x) throw new ArgumentNullException("stop giving functions null idiot");
//instead of
if(x is not null) ...
```

Allow `x is? y` in place of `x is (type of x or T if x is Nullable<T>) y`:
```cs
string? x = null;

if(x is? message) Console.WriteLine(message);
//instead of
if(x is string message) Console.WriteLine(message);
```

Allow `type(many parameter types with optional names)` as a delegate type:
```cs
void DoNothing() {}

void() myFunc = DoNothing;
//instead of
Action myFunc = ...
//or
delegate void MyFunc(); 
MyFunc myFunc = ...

void doubleInt(ref int v) => v *= 2;

void(ref int) m = doubleInt;
//instead of
delegate void MyActionType(ref int v);
MyActionType m = ...


//NOTE: the following is valid, and simply allows for better code suggestions, v is irrelevant to the compiler (given that it is not an invalid identifier)
void(ref int v) m = doubleInt;
```

**Comment by huoyaoyuan** [comment]
Although the usages are common, the proposed syntax isn't straight for users who doesn't read this proposal.

**Reply by dylanrafael05** [reply]
Im unsure what you mean by not "straight"

If you mean straightfoward, then I would say that many other features are not straightforward, such as `??` and `??=`, `?:` conditional statements, or the divergence between `or` and `||`. These syntaxes are designed to require no new keywords and to be relatively unambigious (with the exception of the inline delegate types, which may be ambiguous with a function call if no parameters are provided, but an easy remedy is to prefer function calls in all cases except when a type is expected)

**Comment by HaloFour** [comment]
IMO it's not particularly useful to have a meta-discussion over multiple language feature proposals in one.  It only makes the comments more confusing to follow.

**Reply by dylanrafael05** [reply]
Should I split these into separate discussions? (I would leave this post up as just a suggestion for the first pattern, and others will be placed into separate discussions to limit clutter)

**Reply by theunrepentantgeek** [reply]
Don't go drastically modifying the original post on this discussion - it just would create confusion because the comments already made wouldn't match the revised topic. It's a form of retcon that's frowned upon.

If a feature is worth adopting (and FWIW I do not think any of these are even worth discussing), then it should be a separate discussion. Link each one back to this for context. Once they're created, you can take care of clutter concerns by closing this one.

**Comment by theunrepentantgeek** [comment]
Let's take just the first one of your suggestions:

Allow not x in place of x is not null:

``` csharp
if (not x) 
{
    throw new ArgumentNullException(
        nameof(x), 
        "Stop giving functions null idiot");
}
```

For anyone with recent C# experience, this looks nothing like a pattern match because it's lacking the **is** keyword, so it would be confusing.

For anyone with experience in Delphi, Pascal, Visual Basic, or any of the related languages, this condition looks exactly like `x` must be a **boolean**, with the body of the **if** running if the value is **false**. Given that the type of `x` is not **bool**, this would be confusing.

For anyone with experience in Ruby, Python, JavaScript, TypeScript or any language where truthy and falsey are commonly used idioms, this looks very much like such a test - when it's not, because `if (x)` still wouldn't be valid. They'd be confused too.

True enough that C# does it's own thing - it's not prevented from doing things just because other languages do or don't do them - but creating a language feature expressly to make it more confusing for developers who work across multiple tech stacks seems like an extremely retrograde step.

**Reply by dylanrafael05** [reply]
In that case, would a modification to my second proposal follow C# standards more closely?

```cs
//instead of is? use is not null
if(x is not null y) {
    //y is defined to have the same type as x, or T if x is Nullable<T>
}
```

**Reply by jnm2** [reply]
That's https://github.com/dotnet/csharplang/discussions/4645 and maybe has been mentioned in other threads.
